Fix VertxGrpcExporter reponse status handling

luneo7 opened the pull request:
GRPC status and status message comes from trailers so it needs to be handled in the end handler, if doing before the message if fully sent, trailers are not there yet so there won't be no GRPC status and also no status message.

luneo7 committed:
Fix VertxGrpcExporter reponse status handling

brunobat commented:
Thanks @luneo7. Will checkout the PR and see what happens with the demo apps.

brunobat reviewed (commented):
Do you think you can come up with a test for this @luneo7?
I think this makes total sense. 
@alesj, what do you think?

alesj commented on extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/exporter/otlp/VertxGrpcExporter.java:
```diff
@@ -201,20 +202,39 @@ public void handle(GrpcClientRequest<Buffer, Buffer> request) {
+                                        Level.SEVERE,
+                                        "Failed to export "
+                                                + type
+                                                + "s. The response could not be parsed. Full error message: "
```
How do you know that the response could not be parsed?

luneo7 replied:
Right... actually the stream failed, changed the error log to show that.

alesj commented on extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/exporter/otlp/VertxGrpcExporter.java:
```diff
@@ -201,20 +202,39 @@ public void handle(GrpcClientRequest<Buffer, Buffer> request) {
+                                if (status == GrpcStatus.OK) {
+                                    exporterMetrics.addSuccess(numItems);
+                                    result.succeed();
+                                    return;
```
Perhaps rather do "else"

alesj commented:
Looks good otherwise, with a few cosmetic fixes.
But yeah, a test would be nice / good, if not even mandatory. :-)

Sgitario commented:
Any updates about this pull request? We're experiencing the same issue and confirmed that these changes fixed it.

brunobat commented:
@luneo7 will you complete this PR?

dastrobu commented:
Would be great if we can see a fix for this issue soon.

luneo7 commented:
Will work on this tomorrow, and will check if there is a way to do some testing, as to test this is somehow difficult (it even doesn't have any test right now)...

luneo7 commented:
Checking the existing tests, and also checking OTEL repos, our tests are identical to theirs, even the exporters integration test (https://github.com/open-telemetry/opentelemetry-java/blob/695ed5350b2ef218c38087a627549f4de7b86c22/integration-tests/otlp/src/main/java/io/opentelemetry/integrationtest/OtlpExporterIntegrationTest.java does the same thing that `integration-tests/opentelemetry-vertx-exporter` does). The tests right now get that the implementation is fully working. The thing before it wasn't handling the streaming response correctly, I couldn't figure out an IT to reproduce the original issue, although it is fixed hahaha.

I've addressed the comments, though... if you want to go forward without a new test...

brunobat reviewed (approved):
Thanks very much @luneo7.
Works for me.

brunobat commented:
Should fix: https://github.com/quarkusio/quarkus/issues/35686

famod commented:
Nice, will try this ASAP.

If it fixes the issues I'd strongly suggest backporting it.

famod commented:
This fix should also take effect if cherry-picked on top of the 3.7 branch, shouldn't it?
Thats what I did and I'm still seeing the issue described in #35686. :-/
